File: 백준/Silver/1463. 1로 만들기/1로 만들기.java
import java.io.*;
import java.util.*;


public class Main {
    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        int N = Integer.parseInt(br.readLine());
        Deque<Integer[]> dq = new LinkedList<>();
        dq.addLast(new Integer[] {N, 0});        
        while (true){
            Integer[] nums = dq.pollFirst();
            int num = nums[0];
            int cnt = nums[1];
            
            if (num == 1) {
            	bw.write(cnt+"");
            	break;
            }
            
            if (num%3 == 0) {
            	dq.addLast(new Integer[] {num/3, cnt+1});
            }
            
            if (num%2 == 0) {
            	dq.addLast(new Integer[] {num/2, cnt+1});
            }
            
            dq.addLast(new Integer[] {num-1, cnt+1});
        }
        br.close();
        bw.flush();
        bw.close();
    }
    
    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}

File: 백준/Silver/1463. 1로 만들기/1로 만들기.py
import sys
from collections import deque

input = sys.stdin.readline

N = int(input())

memo = [0 for _ in range(N+1)]
queue = deque([[N, 0]])
while queue:
    num, cnt = queue.popleft()

    if num == 1:
        print(cnt)
        break

    if num % 3 == 0 and not memo[num//3]:
        queue.append([num//3, cnt+1])
        memo[num//3] = cnt+1

    if num % 2 == 0 and not memo[num//2]:
        queue.append([num//2, cnt+1])
        memo[num//2] = cnt + 1

    if not memo[num-1]:
        queue.append([num-1, cnt+1])
        memo[num-1] = cnt + 1